### PR TITLE
Update the Focus widget

### LINKED ISSUE #83023: FocusNode onKey function is not called
## Steps to Reproduce

<details>
  <summary>Minimal code sample</summary>

```
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Shortcuts(
      shortcuts: <LogicalKeySet, Intent>{
        LogicalKeySet(LogicalKeyboardKey.select): const ActivateIntent(),
      },
      child: MaterialApp(
        title: 'Flutter Demo',
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: MyHomePage(title: 'Flutter Demo Home Page'),
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Row(
          children: [
            MainImageBox(key: ValueKey('asdf')),
          ],
        ),
      ),
    );
  }
}

class MainImageBox extends StatefulWidget {
  const MainImageBox({required Key key}) : super(key: key);

  @override
  _MainImageBoxState createState() => _MainImageBoxState();
}

class _MainImageBoxState extends State<MainImageBox> {
  late FocusNode focusNode;

  @override
  void initState() {
    super.initState();
    focusNode = FocusNode(onKey: handleKey);
  }

  bool handleKey(FocusNode focusNode, RawKeyEvent key) {
    print("handleKey function called");
    return false;
  }

  @override
  Widget build(BuildContext context) {
    return RawMaterialButton(
      focusNode: focusNode,
      focusColor: Colors.transparent,
      splashColor: Colors.transparent,
      onPressed: () async {
        print('press');
      },
      child: Container(
        color: Colors.green,
        width: 230,
        padding: EdgeInsets.fromLTRB(5.0, 0.0, 5.0, 0.0),
      ),
    );
  }

  @override
  void dispose() {
    focusNode.dispose();
    super.dispose();
  }
}

```

</details>

1. Run `Minimal code sample`.
2. Move focus around with arrow keys

**Expected results:** 
FocusNode onKey function should be called

**Actual results:** 
onKey function is not called

Works on Flutter 2.0.2 but not 2.2

<details>
  <summary>Logs</summary>

```
➜  onkey flutter analyze
Analyzing onkey...                                                      
No issues found! (ran in 2.9s)
➜  onkey 

```

```
➜  onkey flutter doctor -v
[✓] Flutter (Channel stable, 2.2.0, on macOS 11.3.1 20E241 darwin-x64, locale en-GB)
    • Flutter version 2.2.0 at /Users/snaebjorn/development/flutter
    • Framework revision b22742018b (6 days ago), 2021-05-14 19:12:57 -0700
    • Engine revision a9d88a4d18
    • Dart version 2.13.0

[!] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/snaebjorn/Library/Android/sdk
    • Platform android-S, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5, Build version 12E262
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2018.2.4)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 29.0.4
    • Dart plugin version 182.4323.44

[✓] VS Code (version 1.54.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.20.0

[✓] Connected device (3 available)
    • sdk google atv x86 (mobile) • emulator-5554 • android-x86    • Android 10 (API 29) (emulator)
    • macOS (desktop)             • macos         • darwin-x64     • macOS 11.3.1 20E241 darwin-x64
    • Chrome (web)                • chrome        • web-javascript • Google Chrome 90.0.4430.212
    ! Error: Living Room is not connected. Xcode will continue when Living Room is connected. (code -13)
    ! Error: Living Room is not connected. Xcode will continue when Living Room is connected. (code -13)
    ! Error: Auglysingadeild4thgen is not connected. Xcode will continue when Auglysingadeild4thgen is connected. (code -13)
    ! Error: Snaeji is not connected. Xcode will continue when Snaeji is connected. (code -13)

! Doctor found issues in 1 category.
➜  onkey 

```

</details>

### PR DESCRIPTION
**Pull request by xu-baolin**
Actually this is continued work of #83233, What has changed by this patch:
1, Addresses https://github.com/flutter/flutter/pull/83233#issuecomment-848017935
2, Also assert as @gspencergoog suggest https://github.com/flutter/flutter/pull/83233#issuecomment-880298467
3, Fixes #83023

In addition, I saw that @dkwingsmt  has worked on this widget recently, please take a look at it, thank you very much.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/focus_manager.dart [hunk 1]
```diff
@@ -1088,8 +1088,12 @@ class FocusNode with DiagnosticableTreeMixin, ChangeNotifier {
...
     FocusOnKeyCallback? onKey,
   }) {
+    assert(
+      onKey != null || this.onKey == null,
+      'Overwriting `focusNode.onKey` to null may not be an intentional design.'
+    );
     _context = context;
-    this.onKey = onKey ?? this.onKey;
+    this.onKey = onKey;
     this.onKeyEvent = onKeyEvent ?? this.onKeyEvent;
```
Comment (xu-baolin) [inline packages/flutter/lib/src/widgets/focus_manager.dart:1097]: @dkwingsmt ,Hi, I am not sure how to do with the `onKeyEvent`, what do you think?

Comment (dkwingsmt) [reply packages/flutter/lib/src/widgets/focus_manager.dart:1097]: Behaviors of `onKeyEvent` should be parallel to those of `onKey`, so yes we should change it too.

Comment (dkwingsmt) [reply packages/flutter/lib/src/widgets/focus_manager.dart:1097]: But I think taking `null` as rewriting to `null` is counter-intuitive. If the developer really wants to rewrite the key handler to "not do anything", they can use an empty callback `() => KeyEventResult.ignore`. Having to specify the `onKey` on every call is pretty tedious, considering it's already assignable for many times before that.

Comment (dkwingsmt) [reply packages/flutter/lib/src/widgets/focus_manager.dart:1097]: If we eventually decided to move forward with this change, we need to update the doc too.

**Pull request closed without merging**